### ISSUE TITLE
 Availability of coredns_cache metrics are affected by server block order

### ISSUE THREAD

**Issue by saj** [coredns/coredns#2241]
This is easier for me to explain by walking through a reproduction of unexpected behaviour.  I will walk through two configuration cases.  The first case results in expected behaviour.  The second case, which is only a slight change from the first, results in unexpected behaviour.  Apologies for the length.

---

**Case 1: First server block activates the `cache` plugin**

`Corefile`:

```
.:1053 {
    prometheus 0.0.0.0:9153
    errors

    cache {
        prefetch 2
    }

    forward . 8.8.8.8 {
        max_fails 0
    }
}

.:1054 {
    prometheus 0.0.0.0:9153
    errors

    forward . 8.8.8.8 {
        max_fails 0
    }
}
```

CoreDNS started with:

```
docker run --name coredns --rm \
  --mount type=bind,src=/private/tmp/Corefile,dst=/Corefile,readonly \
  -p 1053:1053/udp -p 1054:1054/udp -p 9153:9153/tcp \
  coredns/coredns:1.2.5
```

Queries to both server blocks works as expected:

```
% dig @127.0.0.1 -p 1053 google.com +noquestion +noadditional +noauthority

; <<>> DiG 9.10.6 <<>> @127.0.0.1 -p 1053 google.com +noquestion +noadditional +noauthority
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 10608
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 4, ADDITIONAL: 9

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; ANSWER SECTION:
google.com.             196     IN      A       216.58.203.110

;; Query time: 13 msec
;; SERVER: 127.0.0.1#1053(127.0.0.1)
;; WHEN: Sat Oct 27 06:35:47 AEDT 2018
;; MSG SIZE  rcvd: 505

% dig @127.0.0.1 -p 1053 google.com +noquestion +noadditional +noauthority

; <<>> DiG 9.10.6 <<>> @127.0.0.1 -p 1053 google.com +noquestion +noadditional +noauthority
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 35353
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 4, ADDITIONAL: 8

;; ANSWER SECTION:
google.com.             194     IN      A       216.58.203.110

;; Query time: 2 msec
;; SERVER: 127.0.0.1#1053(127.0.0.1)
;; WHEN: Sat Oct 27 06:35:50 AEDT 2018
;; MSG SIZE  rcvd: 494
```

(Note the response time on the second query.  Apparent cache hit.)

```
% dig @127.0.0.1 -p 1054 google.com +noquestion +noadditional +noauthority

; <<>> DiG 9.10.6 <<>> @127.0.0.1 -p 1054 google.com +noquestion +noadditional +noauthority
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 11485
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 4, ADDITIONAL: 9

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; ANSWER SECTION:
google.com.             92      IN      A       216.58.203.110

;; Query time: 17 msec
;; SERVER: 127.0.0.1#1054(127.0.0.1)
;; WHEN: Sat Oct 27 06:37:31 AEDT 2018
;; MSG SIZE  rcvd: 505

% dig @127.0.0.1 -p 1054 google.com +noquestion +noadditional +noauthority

; <<>> DiG 9.10.6 <<>> @127.0.0.1 -p 1054 google.com +noquestion +noadditional +noauthority
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 55100
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 4, ADDITIONAL: 9

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; ANSWER SECTION:
google.com.             90      IN      A       216.58.203.110

;; Query time: 27 msec
;; SERVER: 127.0.0.1#1054(127.0.0.1)
;; WHEN: Sat Oct 27 06:37:33 AEDT 2018
;; MSG SIZE  rcvd: 505
```

(Note the response time on the second query.  There is no `cache` on the `.:1054` server block, so this is expected.)

Cache metrics for the `.:1053` server block are available:

```
% curl -s http://localhost:9153/metrics | grep -F coredns_cache_
# HELP coredns_cache_hits_total The count of cache hits.
# TYPE coredns_cache_hits_total counter
coredns_cache_hits_total{server="dns://:1053",type="success"} 1
# HELP coredns_cache_misses_total The count of cache misses.
# TYPE coredns_cache_misses_total counter
coredns_cache_misses_total{server="dns://:1053"} 1
# HELP coredns_cache_size The number of elements in the cache.
# TYPE coredns_cache_size gauge
coredns_cache_size{server="dns://:1053",type="denial"} 0
coredns_cache_size{server="dns://:1053",type="success"} 1
```

All is well so far.

Now, let's flip the server blocks in the `Corefile` and try again.

---

**Case 2: Second server block activates the cache plugin**

`Corefile`:

```
.:1054 {
    prometheus 0.0.0.0:9153
    errors

    forward . 8.8.8.8 {
        max_fails 0
    }
}

.:1053 {
    prometheus 0.0.0.0:9153
    errors

    cache {
        prefetch 2
    }

    forward . 8.8.8.8 {
        max_fails 0
    }
}
```

CoreDNS was started as before.

Queries to both server blocks still works as expected:

```
% dig @127.0.0.1 -p 1053 google.com +noquestion +noadditional +noauthority

; <<>> DiG 9.10.6 <<>> @127.0.0.1 -p 1053 google.com +noquestion +noadditional +noauthority
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 28420
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 4, ADDITIONAL: 9

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; ANSWER SECTION:
google.com.             300     IN      A       216.58.203.110

;; Query time: 104 msec
;; SERVER: 127.0.0.1#1053(127.0.0.1)
;; WHEN: Sat Oct 27 06:44:59 AEDT 2018
;; MSG SIZE  rcvd: 505

% dig @127.0.0.1 -p 1053 google.com +noquestion +noadditional +noauthority

; <<>> DiG 9.10.6 <<>> @127.0.0.1 -p 1053 google.com +noquestion +noadditional +noauthority
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 10923
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 4, ADDITIONAL: 8

;; ANSWER SECTION:
google.com.             299     IN      A       216.58.203.110

;; Query time: 6 msec
;; SERVER: 127.0.0.1#1053(127.0.0.1)
;; WHEN: Sat Oct 27 06:45:01 AEDT 2018
;; MSG SIZE  rcvd: 494
```

(Note the response time on the second query.  Apparent cache hit.)

```
% dig @127.0.0.1 -p 1054 google.com +noquestion +noadditional +noauthority

; <<>> DiG 9.10.6 <<>> @127.0.0.1 -p 1054 google.com +noquestion +noadditional +noauthority
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 63637
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 4, ADDITIONAL: 9

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; ANSWER SECTION:
google.com.             219     IN      A       216.58.203.110

;; Query time: 10 msec
;; SERVER: 127.0.0.1#1054(127.0.0.1)
;; WHEN: Sat Oct 27 06:46:20 AEDT 2018
;; MSG SIZE  rcvd: 505

% dig @127.0.0.1 -p 1054 google.com +noquestion +noadditional +noauthority

; <<>> DiG 9.10.6 <<>> @127.0.0.1 -p 1054 google.com +noquestion +noadditional +noauthority
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 22939
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 4, ADDITIONAL: 9

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; ANSWER SECTION:
google.com.             217     IN      A       216.58.203.110

;; Query time: 9 msec
;; SERVER: 127.0.0.1#1054(127.0.0.1)
;; WHEN: Sat Oct 27 06:46:22 AEDT 2018
;; MSG SIZE  rcvd: 505
```

However, cache metrics for the `.:1053` server block are **no longer available**:

```
% curl -s http://localhost:9153/metrics | grep -F coredns_cache_ | wc -l
       0
```

Here is a list of all the metrics that _were_ available in this case:

```
% curl -s http://localhost:9153/metrics \
  | awk -F '[{} ]' '/^[^#]/ {print $1}' \
  | sort | uniq
coredns_build_info
coredns_dns_request_count_total
coredns_dns_request_duration_seconds_bucket
coredns_dns_request_duration_seconds_count
coredns_dns_request_duration_seconds_sum
coredns_dns_request_size_bytes_bucket
coredns_dns_request_size_bytes_count
coredns_dns_request_size_bytes_sum
coredns_dns_request_type_count_total
coredns_dns_response_rcode_count_total
coredns_dns_response_size_bytes_bucket
coredns_dns_response_size_bytes_count
coredns_dns_response_size_bytes_sum
coredns_forward_request_count_total
coredns_forward_request_duration_seconds_bucket
coredns_forward_request_duration_seconds_count
coredns_forward_request_duration_seconds_sum
coredns_forward_response_rcode_count_total
coredns_forward_sockets_open
coredns_panic_count_total
go_gc_duration_seconds
go_gc_duration_seconds_count
go_gc_duration_seconds_sum
go_goroutines
go_memstats_alloc_bytes
go_memstats_alloc_bytes_total
go_memstats_buck_hash_sys_bytes
go_memstats_frees_total
go_memstats_gc_sys_bytes
go_memstats_heap_alloc_bytes
go_memstats_heap_idle_bytes
go_memstats_heap_inuse_bytes
go_memstats_heap_objects
go_memstats_heap_released_bytes_total
go_memstats_heap_sys_bytes
go_memstats_last_gc_time_seconds
go_memstats_lookups_total
go_memstats_mallocs_total
go_memstats_mcache_inuse_bytes
go_memstats_mcache_sys_bytes
go_memstats_mspan_inuse_bytes
go_memstats_mspan_sys_bytes
go_memstats_next_gc_bytes
go_memstats_other_sys_bytes
go_memstats_stack_inuse_bytes
go_memstats_stack_sys_bytes
go_memstats_sys_bytes
process_cpu_seconds_total
process_max_fds
process_open_fds
process_resident_memory_bytes
process_start_time_seconds
process_virtual_memory_bytes
```

**Comment by fturib** [comment]
Yes ... and I found why.
When several metrics on the same address, only the first one is really active.

And the cache of the second stanza register itself for metrics on the second metrics that is not active.

I am preparing a fix.

**Comment by saj** [comment]
> When several metrics on the same address, only the first one is really active.
> 
> And the cache of the second stanza register itself for metrics on the second metrics that is not active.

By the by, I think the same thing happens with the `forward` plugin.  If all blocks include the `prometheus` plugin, but the first block does not include the `forward` plugin (but a later block does), `forward` metrics will not be served from the HTTP handler.

**Comment by fturib** [comment]
@saj: I agree, does not depend of the plugin itself. it just any plugin that is not if the first block will not have its metric collectors registered properly.

**Comment by fturib** [comment]
fixed with #2246
